### ISSUE TITLE
Question on Converting HumanML3D Format to MotionFix Format

### ISSUE THREAD

**Issue by noah6868** [athn-nik/motionfix#14]
Thank you for your excellent work on MotionFix! I really appreciate the quality of the dataset and the clarity of the paper.

I’m currently trying to convert the HumanML3D dataset format to the one used in MotionFix. Specifically, I have motion sequences in the format``` [T, 22, 3] (joint positions)```, and I’d like to construct the full MotionFix-style motion dictionary for the ```source motion```, which I understand is structured as:
```python

{
  'rots': Tensor of shape [T, 66],          
  'trans': Tensor of shape [T, 3],          
  'joint_positions': Tensor of shape [T, 22, 3]
}
```

Would it be possible for you to share how you compute the 'rots' and 'trans' components from the joint positions? Any code snippets or guidance would be greatly appreciated.

Thanks again for your contribution!

Best regards,
Jiawei

**Comment by noah6868** [comment]
I also noticed that the HumanML3D dataset provides motion data in the``` [T, 263] ```format.
I also attempted to convert this skeletal representation into the MotionFix format, but the process failed, and I'm not sure whether the code I tried is correct.
Here is the code I attempted to run:
```python
mean = np.load(mean_path, allow_pickle=True)
std = np.load(std_path, allow_pickle=True)
npy_0 = np.load(npy_path, allow_pickle=True) # [T, 263]
def convert_t2m_motion_to_motionfix(motion: np.ndarray, mean, std, smpl_layer):
    motion = motion * std + mean  # 还原
    rot6d = motion[:, 67:193].reshape(-1, 21, 6)
    rots_body = transform_body_pose(rot6d, "6d->aa")
    pelvis_rot6d = torch.tensor([[1, 0, 0, 0, 1, 0]], dtype=torch.float32).repeat(len(rots_body), 1)
    pelvis_rot_aa = transform_body_pose(pelvis_rot6d, "6d->aa")
    rots = torch.cat([pelvis_rot_aa.unsqueeze(1), rots_body], dim=1).reshape(len(rots_body), -1)
    trans = torch.from_numpy(motion[:, :3]).float()

    smpl_output = smpl_layer(transl=trans,
                             global_orient=rots[:, :3].reshape(-1, 1, 3),
                             body_pose=rots[:, 3:].reshape(-1, 21, 3))
    joints = smpl_output.joints[:, :22]
    return {
        'rots': rots,
        'trans': trans,
        'joint_positions': joints
    }
motionfix = convert_t2m_motion_to_motionfix(npy_0, mean, std, body_model)
```
I'm not sure whether my approach is correct or if I'm missing something:


<img width="1092" height="270" alt="Image" src="https://github.com/user-attachments/assets/402e20d1-3860-411e-bf46-6cebccbedc18" />

**Comment by atnikos** [comment]
Hey @noah6868 ,
Thanks for the nice words :) 
Q1: You do not need to do that (you can via IK), but is already done in AMASS for you. If you download the original AMASS SMPL-H data they have those parameters that I usee (rotations in axis-angle format, and translation). All you need to do is to map HML3D data sequence names to those in AMASS . HML3D should have the original paths and they should look like this : `CMU/75/75_18_poses`

Q2: Again there is no need to correct or convert as the process they follow is not super straightforward and SMPL-compatible. You can avoid this all together by downloading the original data and using SMPL+H directly from the original amass website.

**Comment by atnikos** [comment]
Please let me know if you are missing any of my points, and feel free to ping me for any further help!

**Comment by noah6868** [comment]
Thanks for the clarification!

Just to clarify my situation: I’m not using the original HML3D dataset, but rather motion data in the same format as HML3D, generated from a model (e.g., in``` [T, 263]``` format). So I don't have access to the original AMASS sequences to retrieve SMPL parameters directly.

Would it be possible to achieve this using inverse kinematics (IK) from ``` [T, 263]``` or ```[T, 22, 3] ``` to motionfix dataset format? If so, how could this be implemented?
**motionfix dataset format:**
```
{
  'rots': Tensor [T, 66],
  'trans': Tensor [T, 3],
  'joint_positions': Tensor [T, 22, 3]
}
```
Or would you suggest a better way to achieve this from generated joint positions?

**Comment by atnikos** [comment]
I see, what about training this model with rotations? which model it is? it shouldn't be hard to achieve this.
You have 2 ways:
- Re-train with parameters that are SMPL compatible without conversions.
- Use this [code](https://github.com/wangsen1312/joints2smpl) to convert 3D positions to SMPL and then you should be good to go. But, please be aware that doing so will lead to your motion containing some artefacts since IK is ill-posed -- i.e., joint positions are not enough to explain the rotation fully. Think about you twisting your forearm (positions do not change, but rotations do). If such artefacts - which can be minimal - do not bother you, you can do the conversion.

**Comment by noah6868** [comment]
I noticed that the HumanML3D dataset also includes the 6D representation of ``` rot_data ```, 

<img width="364" height="253" alt="Image" src="https://github.com/user-attachments/assets/8083be08-9df9-4660-897e-29f7def31495" />

<img width="744" height="484" alt="Image" src="https://github.com/user-attachments/assets/fd516580-d3b9-41c3-823b-f54f37bdf588" />

so I extracted it for use. However, the visualization results were not satisfactory. I would like to understand the``` rots [T,66], trans [T,3] and joint_positions [T,22,3] ```in the MotionFix dataset

<img width="1261" height="146" alt="Image" src="https://github.com/user-attachments/assets/a38c2ac4-e5b5-4432-8165-eb0c1459e45f" />

 — is there any detailed explanation of what these parameters specifically represent? Also, how do they differ from the ```rot_data``` in the HumanML3D dataset’s``` [T,263] ```format?

**Comment by noah6868** [comment]
> I see, what about training this model with rotations? which model it is? it shouldn't be hard to achieve this. You have 2 ways:
> 
>     * Re-train with parameters that are SMPL compatible without conversions.
> 
>     * Use this [code](https://github.com/wangsen1312/joints2smpl) to convert 3D positions to SMPL and then you should be good to go. But, please be aware that doing so will lead to your motion containing some artefacts since IK is ill-posed -- i.e., joint positions are not enough to explain the rotation fully. Think about you twisting your forearm (positions do not change, but rotations do). If such artefacts - which can be minimal - do not bother you, you can do the conversion.

I’ve finished converting the skeleton format, but the conversion is too slow, so I’m still looking for a faster method.
